at_client/src/main/java/org/atsign/client/util/KeyStringUtil.java

```diff
@@ -0,0 +1,134 @@
+        // (7) SelfKey                  == @smoothalligator:lemon@smoothalligator
+        String[] split1 = fullKeyName.split(":");
+        // split1 results
+        // 1 == {"public", "signing_publickey@smoothalligator"} [len 2]
+        // 2 == {"cached", "public", "publickey@denise"} [len 3]
+        // 3 == {"@abbcservicesinc", "shared_key@smoothalligator"} [len 2]
+        // 4 == {"cached", "@smoothalligator", "shared_key@abbcservicesinc"} [len 3]
+        // 5 == {"_latestnotificationid.fourballcorporate9@smoothalligator"} [len 1]
+        // 6 == {"shared_key.wildgreen@smoothalligator"} [len 1]
+        
+
+        // all keys may have a namespace
+        if(fullKeyName.contains(".")) {
+            String[] split2 = fullKeyName.split("\\.");
+            if(split2.length > 1) {
+                String[] split3 = split2[1].split("@");
+                _namespace = split3[0];
+            }
+        }
+
+        if(split1.length > 1) {
+            // must be scenarios 1, 2, 3, 4, 
+
+            // PublicKey check
+            if(split1[0].equals("public") || (split1[0].equals("cached") && split1[1].equals("public"))) {
+                // scenario 1 and 2,, it's a public key!
+                _keyType = KeyType.PUBLIC_KEY;
+            }
+
+            // SharedKey check
+            if(split1[0].startsWith("@") || split1[1].startsWith("@")) {
+                // scenario 3 and 4,, it is a SharedKey!
+                _keyType = KeyType.SHARED_KEY;
+                if(split1[0].startsWith("@")) {
+                    _sharedWith = split1[0].substring(1);
+                } else {
+                    _sharedWith = split1[1].substring(1);
+                }
+            }
+
+            String[] split2 = split1[split1.length-1].split("@");
+            // 1 == {"signing_publickey", "smoothalligator"}
+            // 2 == {"publickey", "denise"}
+            // 3 == {"shared_key", "smoothalligator"}
+            // 4 == {"shared_key", "abbcservicesinc"}
+            _keyName = split2[0];
+            _sharedBy = split2[1];
+
+            // PublicKey and SharedKey can be cacheable!
+            if(split1[0].equals("cached")) {
+                _isCached = true;
+            }
+
+            // _sharedBy == _sharedWith => it's a SelfKey
+            if(_sharedBy.equals(_sharedWith)) {
+                _keyType = KeyType.SELF_KEY;
+            }
+
+        } else {
+            // must be scenarios 5 and 6
+            if(split1[0].startsWith("_")) {
+                _keyType = KeyType.PRIVATE_HIDDEN_KEY;
+            } else {
+                _keyType = KeyType.SELF_KEY;
+            }
+
+            String[] split2 = split1[0].split("@");
+            // 5 == {"_latestnotificationid.fourballcorporate9", "smoothalligator"}
+            // 6 == {"shared_key.wildgreen", "smoothalligator"}
+            _keyName = split2[0];
+            _sharedBy = split2[1];
+        }
+    }
+
+}
```

Comment:
It's probably OK to leave this code here in KeyStringUtil as is. Main thing to move is the stuff that is currently inside the for loop in your _getAtKeys() method